CDS-52164-Custom Webhook Trigger for Non-Harness CI System (#730)
harness/developer-hub: docs/platform/11_Triggers/trigger-pipelines-from-a-third-party-ci-system.md

---
title: Trigger Pipelines from a Third Party CI System
description: Trigger pipelines from a non-Harness CI system.
sidebar_position: 6
---

> You can use a custom webhook trigger when passing an image from a non-Harness CI system. 

Author: Run pipleines from any third-party CI system using custom webhook trigger cURL command.

## Before you begin

Make sure you have: 
* A pipeline with a CD stage
* An artifact in an artifact repository

## Configure runtime input for your artifact tag and image path for a stage.

1. Edit the **Service** of a desired stage in a pipeline.
2. In the **Edit Service** page, edit the **Artifacts** section using the pencil icon.
   
   ![](./static/edit-artifacts.png)

> 3. In **Artifact Location** > **Artifact Details**, select **Runtime input** for **Image Path** and **Tag** fields.

Reviewer: You have to select **Continue** twice to get to the Artifact Location dialog, but that information is missing. Please revise this step to make it clear. 

Also, add "the" before **Image Path**.

   
   ![](./ctatic/../static/edit-artifact-details.png)

> 4. Click **Submit**.

Reviewer: ```suggestion
4. Select **Submit**.
```

   
## Create a custom trigger.

> 1. In your Harness pipeline, click **Triggers**.

Reviewer: ```suggestion
1. In your pipeline, select **Triggers** at the top right portion of the dialog.
```

> 1. In your Harness pipeline, click **Triggers**.

Reviewer: A screenshot with the Triggers option circled would be helpful here because if you've never created a Trigger, the location of the Trigger option is not obvious.

2. Click **New Trigger**.
3. In **Webhook**, select **Custom**.
   
    ![](./static/trigger-deployments-using-custom-triggers-01.png)

> 4. Name the new trigger and click **Continue**.

Reviewer: ```suggestion
4. Name the new trigger, and then select **Continue**.
```

   
    ![](./static/custom-webhook-trigger-configuration.png)

> 5. Specify the **Conditions** for running the pipeline and click **Continue**. 

Reviewer: ```suggestion
5. Specify the **Conditions** for running the pipeline, and then select **Continue**. 
```

> 5. Specify the **Conditions** for running the pipeline and click **Continue**. 

Reviewer: "Specify the conditions" is vague. Can you provide some details to guide the user?

Author: Conditions are not mandarory. Have given an example below for better understanding.

>     For example, execute pipeline if the source/target branch name matches a pattern.

Reviewer: ```suggestion
    For example, execute the pipeline if the source or target branch name matches a pattern.
```

    ![](./static/custom-webhook-trigger-conditions.png)

> 6. In **Pipeline Input**, define payload fields so that Harness can map them to the **Image Path** and **Tag** fields.

Reviewer: You have to select **Continue** to get to the Pipeline Input. Please add that missing step.

Author: mentioned select continue in the previous step

> 6. In **Pipeline Input**, define payload fields so that Harness can map them to the **Image Path** and **Tag** fields.

Reviewer: "define payload fields" is vague. Can you provide details?

    ![](./static/custom-webhook-trigger-piepline-input.png)

> 7. Click **Create Trigger**. 

Reviewer: ```suggestion
7. Select **Create Trigger**. 
```

> ## Trigger the pipeline from an thrid-party CI system.

Reviewer: ```suggestion
## Trigger the pipeline from a third-party CI system
```

> 1. Copy and paste the webhook trigger payload as a cURL command on an editor and edit the body fields.

Reviewer: ```suggestion
1. Locate the trigger you created, select the link in the **Webhook** column, and then select **Copy as cURL Command*. 
2. Paste the command in an editor. Edit the body fields.
```

> 1. Copy and paste the webhook trigger payload as a cURL command on an editor and edit the body fields.

Reviewer: "edit the body fields" is vague. What does the user need to edit and why?

    ![](./static/webhook-trigger-copy-curl-command.png)

    Sample cURL command with image and tag fields:
   
    ```
    curl -X POST -H 'content-type: application/json' -H 'X-Api-Key: sample_api_key' --url 'https://app.harness.io/gateway/pipeline/api/webhook/custom/v2?accountIdentifier=<your_account_identifier>&orgIdentifier=default&projectIdentifier=CD_Docs&pipelineIdentifier=Container&triggerIdentifier=Trigger_to_CI' -d '{"image": "library/nginx", "tag":"stable"}'
    ```

    :::note

>     The `-H 'X-Api-Key: sample_api_key'` parameter is used to authorize custom triggers. You can use [Harness API keys](../4_Role-Based-Access-Control/7-add-and-manage-api-keys.md) with this parameter. This is described below in [Custom trigger authorization using API keys](#custom-trigger-authorization-using-api-keys).

Reviewer: The third sentence says "This is described below in...". Don't user directional indicates, such as below, per the Microsoft Style Guide.[ ](https://learn.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/b/below)

    :::
        
    :::note

>     We recommend that you use VS Code or Sublime Text to edit payload information to avoid format issues.

Reviewer: This information should be in the step where you tell them to paste the command into an editor. It's too late at this point.

    :::

> 2. Run the updated cURL command on Terminal. 

Reviewer: If you made the suggestions I made above, then this should be step 3. 

Also, replace "on Terminal" with "in a Terminal".

    Sample success message:
 
    ```
    {
        "status":"SUCCESS",
        "data":{
            "eventCorrelationId":"<event_correlation_id>",
            "apiUrl":"https://app.harness.io/gateway/pipeline/api/webhook/triggerExecutionDetails/<event_correlation_id>?accountIdentifier=<account_identifier>",
            "uiUrl":"https://app.harness.io/ng/#/account/<account_identifier>/cd/orgs/default/projects/CD_Docs/deployments?pipelineIdentifier=<pipeline_name>&page=0",
            "uiSetupUrl":"https://app.harness.io/ng/#/account/<account_identifier>/cd/orgs/default/projects/CD_Docs/pipelines/<pipeline_name>/pipeline-studio/"
        },
        "metaData":null,
        "correlationId":"<correlation_id>"
    }
    ```
   

> 3. Go to Harness **Deployments** to see the pipeline triggered by your custom webhook trigger.

Reviewer: ```suggestion
3. Go to **Deployments** to see the pipeline triggered by your custom webhook trigger.
```

> 3. Go to Harness **Deployments** to see the pipeline triggered by your custom webhook trigger.

Reviewer: If you made the edits I suggested above, this should be step 4.

---
title: Trigger pipelines from a third-party CI system
description: Trigger pipelines from a non-Harness CI system.
sidebar_position: 6
---

Run pipleines from any third-party CI system using a custom webhook trigger cURL command. 

## Requirements

Make sure you have: 
* A pipeline with a CD stage.
* An artifact in an artifact repository.

## Configure runtime input for your artifact tag and image path for a stage

1. Select the pipeline you want to trigger and select the appropriate stage.
2. On the **Service** tab, select the pencil icon next to the service name.

> 3. On the **Edit Service** dialog, scroll down to the **Artifacts** section and select the pencil icon.

Reviewer: In general, it's best to avoid naming UI elements in instructions, if possible. However, as I was following along as a newbie to the CD UI, I found the UI to be overwhelming with so many pieces and parts that change every time you select something. The UI is quite busy. I suggested that Roshni add the names of UI elements in some cases to act as guides for newer users. For example, it took me a few minutes to realize that "Service" is a tab that I had to select. I also recommended that Roshni add "dialog" after "Edit Service" to provide affirmation to the user that they are in the right place because a new dialog pops up after you select the pencil icon in step 3. Perhaps this is a difference of viewpoint between myself as someone who does not use CD regularly vs Michael who knows everything about the UI. If you are writing only for the most advanced user, then you are not writing to a large segment of your audience.

> 3. On the **Edit Service** dialog, scroll down to the **Artifacts** section and select the pencil icon.

Author: I avoided naming UI elements because our audiences are developers. Also, it is consistent with the style and tone of the existing document. Now, should I or should I not mention ui elements like dialog and page?

  
   ![](./static/edit-artifacts.png)
   
4. Verify if the appropriate artifact repository type is selected, then select **Continue**.
5. Verify if the appropriate artifact connector is selected, then select **Continue**. 
6. In **Artifact Location** > **Artifact Details**, select **Runtime input** for the **Image Path** and **Tag** fields.
   
   ![](./ctatic/../static/edit-artifact-details.png)

7. Select **Submit**.
   
## Create a custom trigger

1. In your pipeline, select **Triggers** at the top right portion of the page.

    ![](./static/pipeline-trigger.png)
   
2. Select **New Trigger**.
3. In **Webhook**, select **Custom**.
   
    ![](./static/trigger-deployments-using-custom-triggers-01.png)

4. Name the new trigger, then select **Continue**.
   
    ![](./static/custom-webhook-trigger-configuration.png)

5. **(Optional)** Specify the **Conditions** for running the pipeline and select **Continue**. 
   
    For example, execute the pipeline if the source or target branch name matches a pattern.

    ![](./static/custom-webhook-trigger-conditions.png)

> 6. In the **Pipeline Input** tab, define payloads for the **Image Path** and **Tag** fields so that Harness can map them later when running the trigger.

Reviewer 2: ```suggestion
6. In **Pipeline Input**, define payloads for **Image Path** and **Tag** so that Harness can map them later when running the trigger.
```

    ![](./static/custom-webhook-trigger-piepline-input.png)

7. Select **Create Trigger**. 
   
## Trigger the pipeline from a thrid-party CI system

> 1. Locate the trigger you created, select the link in the **Webhook** column and select **Copy as cURL Command**.

Reviewer 2: ```suggestion
1. In your pipeline **Triggers**, locate the trigger you created, select the link in the **Webhook** column, and select **Copy as cURL Command**.
```

    ```

    The `-H 'X-Api-Key: sample_api_key'` parameter is used to authorize custom triggers. You can use [Harness API keys](../4_Role-Based-Access-Control/7-add-and-manage-api-keys.md) with this parameter. For more information, go to [Custom trigger authorization using API keys](#custom-trigger-authorization-using-api-keys).
    

3. Run the updated cURL command in a Terminal. 

    Sample success message:
 
    ```
    {
        "status":"SUCCESS",
        "data":{
            "eventCorrelationId":"<event_correlation_id>",
            "apiUrl":"https://app.harness.io/gateway/pipeline/api/webhook/triggerExecutionDetails/<event_correlation_id>?accountIdentifier=<account_identifier>",
            "uiUrl":"https://app.harness.io/ng/#/account/<account_identifier>/cd/orgs/default/projects/CD_Docs/deployments?pipelineIdentifier=<pipeline_name>&page=0",
            "uiSetupUrl":"https://app.harness.io/ng/#/account/<account_identifier>/cd/orgs/default/projects/CD_Docs/pipelines/<pipeline_name>/pipeline-studio/"
        },
        "metaData":null,
        "correlationId":"<correlation_id>"
    }
    ```
   

> 4. Go to **Deployments** to see the pipeline triggered by your custom webhook trigger.

Reviewer 2: ```suggestion
4. In Harness, go to **Deployments** or the pipeline's **Execution History** to see the pipeline triggered by your custom webhook trigger.
```
